JENKINS-60730 add new trait to filter out Branches and Tags older than X

dsteiert opened the pull request:
A new trait called BranchOrTagAgeDiscoveryTrait is introduced to filter out branches and tags that are older than `maxDaysOld`. The `maxDaysOld` parameter can be set in the UI when the trait is selected. All branches and tags older than the specified amount of days are ignored after that.

* JENKINS issue(s):
    * https://issues.jenkins-ci.org/browse/JENKINS-60730
* Description:
    * A new trait called BranchOrTagAgeDiscoveryTrait is introduced to filter out branches and tags that are older than `maxDaysOld`. The `maxDaysOld` parameter can be set in the UI when the trait is selected. All branches and tags older than the specified amount of days are ignored after that.
* Documentation changes:
    * Link to related jenkins.io PR or explanation for why doc change not needed
* Users/aliases to notify:
    * ...

bitwiseman commented:
To be clear this is branches or tags whose head commit has a commit date older than X.  It is possible to create a branch or tag today which pointing to a commit from 10 days ago, which would be ignored if your window were 5 days.

dsteiert commented:
Thank you for your observation! I updated the PR to include a help-message making it abundantly clear what the `maxDaysOld` parameter is for and how it works.

dsteiert commented:
I also refactored the code to use `new Date().getTime()` now instead of `System.currentTimeMillis()` and added a fallback in case `maxDaysOld` or the `lastModified` timestamp are `0`. In that case the default behaviour takes over and the branch/tag is not excluded.

bitwiseman commented on src/main/resources/org/jenkinsci/plugins/github_branch_source/BranchOrTagAgeDiscoveryTrait/help-maxDaysOld.html:
```diff
@@ -0,0 +1,35 @@
+    <p>
+    	Some SCM plugins do not support the retrieval of timestamps of commits. In this
+    	case the time-restriction will be ignored. Known plugins in which this is the case 
+    	are the <a href="https://github.com/jenkinsci/mercurial-plugin">mercurial plugin</a>
```
Wait.  Does this mean that you expect this trait to work on mercurial?  But GitHub doesn't do mercurial repos.

dsteiert replied:
You are right, sorry about that. I got a bit carried away. I also have a pull-request for another plugin in which this was relevant and I forgot to remove/change this part. I removed it, since it does not apply to this plugin.

dsteiert commented:
I also found a mistake in the `messages.properties` where it stated before: `Discover branches or tags older than a certain age` to say `Discover branches or tags newer than X days`, which I think is more accurate.

bitwiseman commented on src/main/resources/org/jenkinsci/plugins/github_branch_source/Messages.properties:
```diff
@@ -52,6 +52,8 @@ BranchDiscoveryTrait.onlyPRs=Only branches that are also filed as PRs
 BranchDiscoveryTrait.authorityDisplayName=Trust origin branches
 BranchDiscoveryTrait.displayName=Discover branches
+BranchOrTagAgeDiscoveryTrait.displayName=Discover branches or tags newer than X days
+BranchOrTagAgeDiscoveryTrait.authorityDisplayName=Trust branches or tags newer than X days
```
Is this correct? (Sorry, for the multiple cycles of comments I thought I'd added this one yesterday, but hadn't clicked the button.)

dsteiert replied:
That is the location I changed. It is correct now. Any branches or tags not older than `X` days  are accepted.

bitwiseman commented on src/main/java/org/jenkinsci/plugins/github_branch_source/BranchOrTagAgeDiscoveryTrait.java:
```diff
@@ -0,0 +1,245 @@
+        @Restricted(NoExternalUse.class)
+        public FormValidation doCheckMaxDaysOld(@QueryParameter int maxDaysOld) {
+            if (maxDaysOld <= 0) {
+                return FormValidation.warning("The maximum age of a commit should be greater than 0. Otherwise no branches are found");
```
This has changed now right?

dsteiert replied:
Updated.

bitwiseman commented on src/main/java/org/jenkinsci/plugins/github_branch_source/BranchOrTagAgeDiscoveryTrait.java:
```diff
@@ -0,0 +1,245 @@
+
+            long daysOld = TimeUnit.MILLISECONDS.toDays(new Date().getTime() - lastModified);
+            if (lastModified != 0 && 
+                    maxDaysOld != 0 && 
```
If user puts in a negative number, we warn but don't stop them - treat it as 0.
```suggestion
                    maxDaysOld <= 0 && 
```

bitwiseman replied:
```suggestion
                    maxDaysOld > 0 && 
```

bitwiseman commented on src/main/java/org/jenkinsci/plugins/github_branch_source/BranchOrTagAgeDiscoveryTrait.java:
```diff
@@ -0,0 +1,245 @@
+/**
+ * A {@link Discovery} trait for GitHub that will discover branches with a certain age on the repository.
+ *
+ * @since 2.2.0
```
With this addition we'll bump the version. 
```suggestion
 * @since 2.6.0
```

bitwiseman commented on src/main/resources/org/jenkinsci/plugins/github_branch_source/BranchOrTagAgeDiscoveryTrait/help-maxDaysOld.html:
```diff
@@ -0,0 +1,25 @@
+<div>
+    <p>
+        The maximum amount of days a commit at the head of a branch, tag or pull request
+        may be old before being ignored. 
```
```suggestion
```

bitwiseman commented on src/main/resources/org/jenkinsci/plugins/github_branch_source/BranchOrTagAgeDiscoveryTrait/help-maxDaysOld.html:
```diff
@@ -0,0 +1,25 @@
+<div>
+    <p>
+        The maximum amount of days a commit at the head of a branch, tag or pull request
```
```suggestion
        The maximum age in days of the head commit of a branch, tags, or pull request.
```

bitwiseman commented on src/main/resources/org/jenkinsci/plugins/github_branch_source/BranchOrTagAgeDiscoveryTrait/help-maxDaysOld.html:
```diff
@@ -0,0 +1,25 @@
+        may be old before being ignored. 
+    </p>
+    <p>
+    	Branches, tags or pull requests older than this will be ignored.
```
```suggestion
    	Branches, tags or pull requests will be ignored if their head is older than this number of days.
```

bitwiseman commented on src/main/resources/org/jenkinsci/plugins/github_branch_source/BranchOrTagAgeDiscoveryTrait/help-maxDaysOld.html:
```diff
@@ -0,0 +1,25 @@
+    <p>
+    	As a basis of the time-restriction, the last modification timestamp of the head commit 
+    	is used together with the current system time. If the current system time is 
+    	2020-01-10 11:20 AM GMT and the maximum commit age is set to 3 days, anything after
```
```suggestion
    	2020-01-10 11:20 AM GMT and the maximum commit age is set to 3 days, only branches, tags or pull requests more recent than
```

bitwiseman commented on src/main/java/org/jenkinsci/plugins/github_branch_source/BranchOrTagAgeDiscoveryTrait.java:
```diff
@@ -0,0 +1,245 @@
+    @Override
+    protected void decorateContext(SCMSourceContext<?, ?> context) {
+        GitHubSCMSourceContext ctx = (GitHubSCMSourceContext) context;
+        ctx.withAuthority(new BranchOrTagAgeSCMHeadAuthority());
```
What does this do?

dsteiert replied:
As far as I could tell it is a way to activate trust for the head and request that is shared to other authorities (according to [this](https://github.com/jenkinsci/scm-api-plugin/blob/master/src/main/java/jenkins/scm/api/trait/SCMHeadAuthority.java#L41-L48)). But you are right, it should probably take `SCMHead` in the Authorities' `checkTrusted` method since we don't just accept branches but also . I will update it.

dsteiert replied:
Let me know if it is not needed from your perspective and I will remove it.

dsteiert commented:
> But before you do other significant work on this, I just realized there might be a problem. I think making this a "Discovery" trait means that branches, tags, and PRs which exceed the max age will not only not be built but will also _not be visible_. That would mean that the job for that item would be removed from a multibranch job tree.
> 
Branches that are too old don't show up, correct. It is not a must for us to see old branches.
> I don't think that is the behavior you want, right? Have you tested this manually to see what happens when a branch was visible but then ages out? (Luckily we've already found an easy way to force this to happen - force pushing an old commit.)
I have tested it and when a branch becomes older than accepted it is greyed out, the text is marked with strikethrough and it is eventually removed. It is not directly wanted behaviour but behaviour we can live with as a trade-off to getting this merged within 2020 compared to having it never merged when we try to create a PR for other plugins. However, I will spend some time to find a way to have jobs visible/usable while not automatically built/active. It might mean making changes in other plugins in addition. 

Nonetheless, having this strategy in the `github-branch-source` plugin would still be required to reduce our pain within the time until there is a solution that is less strict (removing outdated branches - I will update the helpfile to reflect this behaviour).

I will start implementing the change-requests you have mentioned and will address your questions one-by-one while looking into less-restrictive alternatives. It would be great if we can merge this into the `github-branch-source` plugin at some point nonetheless.

dsteiert commented:
Added all your suggestions so far aside from
> Also we'll need more tests.
>
Working on this now.

darxriggs commented:
I haven't had a look in detail at the code but maybe you are also interested in another existing plugin regarding this - see https://github.com/jenkinsci/scm-filter-aged-refs-plugin

dsteiert commented:
@bitwiseman, I added the requested tests. They were already finished last week Friday and I refactored them slightly before pushing. Unfortunately, I guess the work became obsolete with your correct assessment @darxriggs. 
Tests with the `scm-filter-aged-refs-plugin` show that it does exactly what we want aside from filtering out tags. That might be good enough for us atm.
We also will probably spend time on adding a branch strategy to the {{github-branch-source}} plugin or adding a plugin that extends the {{github-branch-source}} plugin's strategies in the future. That way "old" branches could be displayed and manually built, while not automatically being triggered by SCM events. 

However, I think we should still proceed with merging this new trait into the {{github-branch-source}} plugin. It supports Branches, Tags and Pull Requests, has more extensive documentation, doesn't require a second plugin in addition to the {{github-branch-source}} plugin be installed while getting less features (no offence to your mentioned plugin @darxriggs, I just believe there is still enough reason to continue with the PR review).

What do you think @bitwiseman and @darxriggs?

bitwiseman commented:
> Branches that are too old don't show up, correct. It is not a must for us to see old branches.


> I have tested it and when a branch becomes older than accepted it is greyed out, the text is marked with strikethrough and it is eventually removed. It is not directly wanted behaviour but behaviour we can live with as a trade-off to getting this merged within 2020 compared to having it never merged when we try to create a PR for other plugins. However, I will spend some time to find a way to have jobs visible/usable while not automatically built/active. It might mean making changes in other plugins in addition.

I understand it is not a requirement for you but having branches go missing by age is not great. 
Unfortunately, with the default configuration, those will delete immediately.  

I think the behavior your want is based on the `BranchBuildStrategy`, like this: 
https://github.com/jenkinsci/basic-branch-build-strategies-plugin/blob/master/src/main/java/jenkins/branch/buildstrategies/basic/ChangeRequestBuildStrategyImpl.java

To be clear, I'm fine with you checking the change into this plugin, I'm only pointing to that repo so you can see what classes to use.  I think you can make the changes without making changes to 

> Nonetheless, having this strategy in the `github-branch-source` plugin would still be required to reduce our pain within the time until there is a solution that is less strict (removing outdated branches - I will update the helpfile to reflect this behaviour).

To repeat, I'm fine with you checking the change into this plugin. 

> I will start implementing the change-requests you have mentioned and will address your questions one-by-one while looking into less-restrictive alternatives. It would be great if we can merge this into the `github-branch-source` plugin at some point nonetheless.

bitwiseman commented:
Finally, thanks for contributing.  I look forward to merging this feature when it is done, it will definitely be of use to a lot of people.

darxriggs commented:
I just wanted to let you know that there is already another plugin. The question is always if code can be shared and a strategy be implemented that can be used by multiple plugins or to duplicate nearly the same code in multiple plugins.

dsteiert commented:
> I think the behavior your want is based on the `BranchBuildStrategy` from branch-api.
> https://github.com/jenkinsci/branch-api-plugin/blob/master/src/main/java/jenkins/branch/BranchBuildStrategy.java
> 
> It would end up looking something like this:
> https://github.com/jenkinsci/basic-branch-build-strategies-plugin/blob/master/src/main/java/jenkins/branch/buildstrategies/basic/ChangeRequestBuildStrategyImpl.java
> 

You are right, I had looked at the `basic-branch-build-strategies` plugin before but hadn't seen what I needed/wanted. After a lot more research and digging we got one of the strategies, `skipInitialBuildOnFirstBranchIndexing` working also with the Job-DSL. Together with the functionality of the `github-scm-filter-aged-refs` plugin and the `TagBuildStrategy` of the `basic-branch-build-strategies` plugin, we can probably solve our issues. 
This basically enables us to not build every single branch when the Jenkins is recreated or when there is data corruption and we can filter branches and tags out that are too old. 

Thanks for all your help and the time you took to review and look for solutions! I think we could close this pull request with the other solutions in place.

bitwiseman commented:
@dsteiert Would you be up for writing a blog post for jenkins.io about your configuration?  It sounds interesting.

dsteiert commented:
> @dsteiert Would you be up for writing a blog post for jenkins.io about your configuration? It sounds interesting. I hope I didn't dissuade you from contributing with my feedback.

You didn't dissuade me from contributing. I also prefer the solution of not having to modify existing code that many other teams/companies use every day 😉.

What kind of blog post were you thinking about? 

- The pipeline setup we are working on in general with generic, configurable pipelines for teams (centrally managed pipeline as code)
- This specific problem and the solution we came up with to solve it (the issue with data-loss and not spamming Jenkins with every branch build)

The second one seems almost trivial and embarrassing now after solving it, since I should have seen these strategies and traits but might enable others to more quickly solve this problem. It might be of value then, since I wasn't able to find a solution to this anywhere until now (and we have had this problem already 4 years ago with our older-generation Kubernetes based Microservices pipelines that would kill the cluster due to this.

Let me know, generally I am open to it.

bitwiseman commented:
@dsteiert  
I was thinking of the second one. It may seem simple to you now, but you listed four different plugins interacting and since it took some work for you to figure out it is something others would find helpful. Worth writing a few paragraphs and a screenshot or two about.  

If you want to talk about the first point, that also may be of interest too.  Say hi to the folks in https://gitter.im/jenkinsci/advocacy-and-outreach-sig .

secbug commented:
I know this is a year and a half later, but I'd still like to see something similar implemented here. As of 3+ years ago, scm-filter-aged-refs-plugin seems dead and it would be amazing to have it included in the plugin we've already got anyways. (Plugin dependency is great and all, but multi-level plugin dependency we find to be rough in corporate Jenkins management land)

bitwiseman commented:
@secbug Contribution are welcome, though admittedly this particular problem would be hard for a first time contributor since there isn't one clear right way to do move forward.  Still, maybe you can start a discussion on the `jenkins-dev` email list and see if you can find some collaborators.
